confmap/provider/s3provider/provider.go

```diff
@@ -0,0 +1,139 @@
+
+// New returns a new confmap.Provider that reads the configuration from a file.
+//
+// This Provider supports "s3" scheme, and can be called with a "uri" that follows:
+//   s3-uri : s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
+//
+// One example for s3-uri be like: s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg
+//
+// Examples:
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	return &provider{}
+}
+
+func (fmp *provider) Retrieve(ctx context.Context, uri string, _ confmap.WatcherFunc) (confmap.Retrieved, error) {
+	if !strings.HasPrefix(uri, schemeName+":") {
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not supported by %q provider", uri, schemeName)
+	}
+
+	// Check if users set up their env for S3 Auth check yet
+	if os.Getenv("AWS_ACCESS_KEY_ID") == "" || os.Getenv("AWS_SECRET_ACCESS_KEY") == "" {
+		return confmap.Retrieved{}, fmt.Errorf("unable to fetch access keys for S3 Auth")
+	}
+
+	// Split the uri and get [BUCKET], [REGION], [KEY]
+	bucket, region, key, err := S3URISplit(uri)
+	if err != nil {
+		return confmap.Retrieved{}, fmt.Errorf("%q uri is not valid s3-url", uri)
+	}
+
+	// AWS SDK default config
+	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(region))
+	if err != nil {
+		return confmap.Retrieved{}, fmt.Errorf("AWS SDK's default configuration fail to load")
+	}
+
+	// to create a s3 client and also a s3 downloader
+	// s3 client provides interfaces for Bucket/File Management in Amazon S3
+	// s3 downloader is especially for s3 downloading operation
+	client := s3.NewFromConfig(cfg)
+	downloader := manager.NewDownloader(client)
+
+	// headObject : metadata, to check the length of the YAML bytes
+	headObject, err := client.HeadObject(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return confmap.Retrieved{}, fmt.Errorf("HeadObject failed to fetch : Bucket %q, Key %q, Region %q", bucket, key, region)
+	}
+	buf := make([]byte, int(headObject.ContentLength))
+	w := manager.NewWriteAtBuffer(buf)
+
+	// download op
+	_, err = downloader.Download(ctx, w, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return confmap.Retrieved{}, fmt.Errorf("file in S3 failed to fetch : uri %q", uri)
+	}
+
+	return internal.NewRetrievedFromYAML(w.Bytes())
+}
+
+func (*provider) Scheme() string {
+	return schemeName
+}
+
+func (*provider) Shutdown(context.Context) error {
+	return nil
+}
+
+// S3URISplit splits the s3 uri and get the [BUCKET], [REGION], [KEY] in it
+// INPUT : s3 uri (like s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY])
+// OUTPUT :
+//		-  [BUCKET] : The name of a bucket in Amazon S3.
+//		-  [REGION] : Where are servers from, e.g. us-west-2.
+//		-  [KEY]    : The key exists in a given bucket, can be used to retrieve a file.
+func S3URISplit(uri string) (string, string, string, error) {
```

Comment:
This should not be exported since it is only useful as an internal helper.  

```suggestion
func s3URISplit(uri string) (string, string, string, error) {
```

Reply:
Thanks for the clarification! Pushed new codes for this comment yet! :)